### PR TITLE
fix issue #6024 with ThreadLocalStoragePointer

### PR DESCRIPTION
**Pull request by cr4zyserb**

### CODE REVIEW

**Comment by derekbruening** [comment]
Please request a review when ready: looks like this is still in-progress I assume.

**Comment by cr4zyserb** [comment]
No, its done. I've only fixed some formatting issues to match what is expected to pass checks.

**Review by derekbruening — COMMENTED** [review]
Thank you for submitting a fix.  As mentioned, we have some conventions on the issue title, etc.  The gist of your fix makes sense but I had a question -- see the comments.

File: core/win32/loader.c [hunk 1]
```diff
@@ -253,13 +253,21 @@ os_loader_init_prologue(void)
...
-                "initial thread TEB->ThreadLocalStoragePointer=" PFX "\n",
-                pre_static_tls);
-        }
+        /* FIX: i#6024: when kernelbase.dll has TLS or client which will lead to
+         * crash. Solution is to always save pre_static_tls but do not set 
+         * ThreadLocalStoragePointer to NULL. If we set it to NULL, it will lead to
+         * crash when application uses ThreadLocalStoragePointer (eg. __declspec(thread)
+         * is used), but client or kernelbase.dll (Win11/Win22H2) or client doesn't
+         * have TLS which will fail to restore TheadLocalStoragePointer later in
+         * calls to should_swap_teb_static_tls() as tlx_next_idx is 0
```
[the remark this answers is not in the document]
Comment (cr4zyserb) [reply core/win32/loader.c:262]: Correct, this is happening for the 1st thread.  As far as I was debugging and tracking this issue ``context->priv_static_tls`` is initialized in ``os_loader_thread_init_prologue`` and is not affected. I think error was in ``should_swap_teb_static_tls`` being called before the time to check and save TLS for the 1st thread, when there is no client loaded yet, thus would be always NULL.

Now issue doesn't exist if client (or dll dependency) doesn't use TLS,  ``should_swap_teb_static_tls`` will always return false, and TLS stays as it is. Problem only comes when there is TLS isolation:

```os_loader_init_prologue``` -> ```should_swap_teb_static_tls``` returns **false**, now we enter into ``privload_os_finalize`` and ``tls_next_idx`` is incremented but ``pre_static_tls`` is still NULL, and once we hit ``os_loader_thread_init_prologue`` due to the call to ``should_swap_teb_static_tls`` will set TLS to NULL, leading program to crash if it uses TLS (in my case it was gdi32.dll)

I will go over code guidelines, as provided by @lihasgupta , but that will have to wait for Monday.

[the remark this answers is not in the document]
Comment (cr4zyserb) [reply core/win32/loader.c:262]: You may close this pull request, as I've created a new one with proposed workflow (branch naming, comments removed, and commit having description as mention by @lihasgupta )

**Pull request closed without merging**